app/src/main/java/de/dhbw/corona_world_app/api/APIManager.java

```diff
@@ -1,150 +1,156 @@
-                } else {
-                    attachString = isoCountry.toString();
+        if (countryList.size() <= MAX_COUNTRY_LIST_SIZE) {
+            for (ISOCountry isoCountry : countryList) {
+
+                //make api-call
+                Future<String> future = service.submit(() -> {
+                    String url = API.HEROKU.getUrl();
+                    url += API.HEROKU.getOneCountry();
+
+                    String attachString;
+                    if (Mapper.isInReverseMap(API.HEROKU, isoCountry)) {
+                        attachString = Mapper.mapISOCountryToName(API.HEROKU, isoCountry);
+                    } else {
+                        attachString = isoCountry.toString();
+                    }
+                    url += attachString;
+                    return createAPICall(url);
                 }
-                url += "/" + attachString;
-            }
-
-            //make api-call
-            String apiReturn = createAPICall(url);
-
-            //parse api-return into country and return
-            Country country = new Country(isoCountry.toString());
-            StringToCountryParser.parseFromHeroOneCountry(apiReturn, country);
-
-            //check if popCount is to be shown
-            if (criteriaList.contains(Criteria.POPULATION)) {
-                StringToCountryParser.parsePopCount(createAPICall("https://restcountries.eu/rest/v2/name/" + isoCountry.getISOCode() + "?fullText=true"), country);
+                );
+                Country country = new Country(isoCountry.toString());
+                try {
+                    String apiReturn = future.get();
+                    //parse api-return into country and return
+
+                    StringToCountryParser.parseFromHeroOneCountry(apiReturn, country);
+
+                    //check if popCount is to be shown
+                    if (criteriaList.contains(Criteria.POPULATION)) {
+                        Future<String> future2 = service.submit(() -> createAPICall(API.RESTCOUNTRIES.getUrl() + API.RESTCOUNTRIES.getOneCountry() + isoCountry.getISOCode()));
+                        String apiReturn2 = future2.get();
+                        StringToCountryParser.parsePopCount(apiReturn2, country);
+                    }
+                } catch (ExecutionException e) {
+                    Logger.logE("getData", "Error executing async call\n" + Arrays.toString(e.getStackTrace()));
+                    throw (IOException) Objects.requireNonNull(e.getCause());
+                } catch (InterruptedException e) {
+                    Logger.logE("getData", "Interruption error\n" + Arrays.toString(e.getStackTrace()));
+                }
+                returnList.add(country);
             }
-            returnList.add(country);
+        } else {
+            throw new IllegalArgumentException("Input country list is too big, max allowed="+MAX_COUNTRY_LIST_SIZE);
         }
```

Comment:
Du solltest es nach der For Loop die Resultate des Api-Calls holen, da ansonsten die Api-Calls synchron nacheinander abgefragt werden.

Reply:
Siehe vorheriger Kommentar